src/openms/source/ANALYSIS/ID/SearchDatabase.cpp

```diff
@@ -0,0 +1,337 @@
+
+  if (entries.size() == 0) return;
+
+  all_peptides_ = generate_peptides_(entries);
+
+  all_fragments_ = generate_fragments_();
+  
+  bucketsize_ = size_t(sqrt(all_fragments_.size())); //calculating bucketsize for balanced tree (performance)
+
+  for (size_t i = 0; i < all_fragments_.size(); i += bucketsize_)
+  {
+    bucket_frags_mz_.emplace_back(all_fragments_[i].fragment_mz_);
+  }
+  // sorting the fragments of each bucket by the precursor-m/z
+  #pragma omp parallel for  
+  for (size_t i = 0; i < all_fragments_.size(); i += bucketsize_)
+  { 
+    auto bucket_begin = all_fragments_.begin()+i;
+    auto condition = distance(all_fragments_.begin(), bucket_begin+bucketsize_) >= int(all_fragments_.size());
+
+    sort(bucket_begin, condition ? all_fragments_.end() : bucket_begin+bucketsize_, 
+    [&](const SearchDatabase::Fragment_& l, const SearchDatabase::Fragment_& r) -> bool 
+    {return (all_peptides_[l.peptide_index_].peptide_mz_ < all_peptides_[r.peptide_index_].peptide_mz_);});
+  }
+  
+}
+
+void SearchDatabase::updateMembers_()
+{
+  digestor_enzyme_ = param_.getValue("digestor_enzyme").toString();
+  missed_cleavages_ = param_.getValue("missed_cleavages");
+  peptide_min_mass_ = param_.getValue("peptide_min_mass");
+  peptide_max_mass_ = param_.getValue("peptide_max_mass");
+  peptide_min_length_ = param_.getValue("peptide_min_length");
+  peptide_max_length_ = param_.getValue("peptide_max_length");
+  fragment_min_mz_ = param_.getValue("fragment_min_mz");
+  fragment_max_mz_ = param_.getValue("fragment_max_mz");
+  precursor_mz_tolerance_ = param_.getValue("precursor_mz_tolerance");
+  fragment_mz_tolerance_ = param_.getValue("fragment_mz_tolerance");
+  precursor_mz_tolerance_unit_ = param_.getValue("precursor_mz_tolerance_unit").toString();
+  fragment_mz_tolerance_unit_ = param_.getValue("fragment_mz_tolerance_unit").toString();
+  modifications_fixed_ = ListUtils::toStringList<std::string>(param_.getValue("modifications_fixed"));
+  modifications_variable_ = ListUtils::toStringList<std::string>(param_.getValue("modifications_variable"));
+  max_variable_mods_per_peptide_ = param_.getValue("max_variable_mods_per_peptide");
+}
+
+void SearchDatabase::search(MSSpectrum& spectrum, std::vector<SearchDatabase::Candidate>& candidates) const
+{ 
+  candidates.clear();
+  if (spectrum.size() == 0) return;
+  
+  unordered_set<size_t> index_hash; // saving every candidate only once
+
+  std::vector<Precursor> precursor = spectrum.getPrecursors();
+
+  if (precursor.size() != 1) return;
+
+  double prec_mz = precursor[0].getUnchargedMass();
+
+  spectrum.sortByIntensity(true);
+  size_t count_found = 0;
+  size_t count_rounds = 0;
+
+  auto prec_unit_da = precursor_mz_tolerance_unit_ == UNIT_DA;
+  auto frag_unit_da = fragment_mz_tolerance_unit_ == UNIT_DA;
+
+  for (const auto& peak : spectrum)
+  {
+    if (count_found == 0 && count_rounds == 5) return; // if peak with the 5 highest intensity don't match, skip whole spectrum
```

Comment:
we could think about parameterizing max_missed_rounds.

Reply:
Interesting. Is this heuristic also used by sage? My first impression was that this could lead to decreased sensitivity in case of high intensity precursors + neural losses as well as dominant Imodium ions.

Reply:
I'm not using this heuristic - would definitely reduce sensitivity for high intensity precursors, high intensity labile modifications (chemical probes...), etc.